### DISCUSSION TITLE
add Comment/Notes Functionality

### DISCUSSION THREAD

**Discussion by phipno** [j-kaltes/Juggluco#139] [Ideas]
I wanna be able to add Comments/Notes like the already functionality to set insulin dose, carbon hydrates amount, dextrose... 
Comments are a great way to note down foods or explain certain scenarios to your doctor.

Till now i don't see any way to input raw text into one of the "Etiketten" sadly. 
One more Feature could be that the export/sync to LibreView is possible.

I got some knowledge coding in c And c++ but never worked on a project of this scale. 

I'm going to give it a try though ~ gonna fork this beauty of a baby.

Maybe someone can offer me help and give me a run down of the code, since I think this shouldn't be so hard to implement and would be a nice first-issue.
![IMG_20240315_135511~2](https://github.com/j-kaltes/Juggluco/assets/98472944/e11fa4f9-8081-4ffc-9e6a-67fa86b3640f)

**Comment by j-kaltes** [comment]
![meal](https://github.com/j-kaltes/Juggluco/assets/72013079/072c051f-2fbf-4b85-9614-018ee1891d6e)
![appel](https://github.com/j-kaltes/Juggluco/assets/72013079/b395dbbb-622c-4d81-927c-7a17b57ae74d)

Adding arbitrary notes is impossible. “note down foods” can easily be done. If in left middle menu→”New Amount”, you select “Carbohydrate”, a meal button will become visible in the middle of the dialog.  By pressing it, you can specify a meal out of its component parts, see https://www.juggluco.nl/Jugglucohelp/mealhelp.html

Writing letters to your doctor, you can’t do in Juggluco. But you can create labels for certain events and specify to what degree they were present. See: https://www.juggluco.nl/Jugglucohelp/labelhelp.html

**Comment by thegrobi** [comment]
Notes would be really helpful, not only for docs, but for the patient himself.

**Comment by damgot** [comment]
I'm also looking for this feature.

It would be great to be able to add a note to the graph (which wouldn't be pre-saved as a label), or a ‘note’ label which would accept text and not just numbers, and whose value (text) input would be sent to libreview as a comment.

That's what I'm missing most with juggluco at the moment, the ability to quickly add a comment explaining a particular situation on the graph (and which can be exported to LibreView so that my diabetologist can see this same explanation on his graph).

As these notes can be extremely varied, and not necessarily linked to food, having to pre-create labels doesn't seem to me to be a good solution.

Examples of notes would be : 
- Stress
- Sick
- Heatwave
- Party
- Bad night
- Hypoglycaemia not felt
- Meal delayed
- Air travel
- Hospitalisation
- Fasting
- ...

**Comment by j-kaltes** [comment]
Notes that are of use are not unique, but can repeat itself. Most can be present in a variable degree. 
Juggluco supports up to 40 labels. You mentioned only 10 types.

**Reply by damgot** [reply]
Yes, of course, these are just examples, I could add many more.
Indeed, recurring events are worth adding as labels.
However, we need to be careful about what we create in labels because we can't simply delete a label. If we have a list of 40 labels, we can't delete the 10th one that we no longer need (for example) to make room. We can only delete the last one. We could rename the 10th label (still an example), but in this case all previous notes encoded using this 10th label would be renamed too. Which is not practical.

Note that this isn't a criticism at all, but just a suggestion for a new feature, which would be very useful to me with the way I use Juggluco (which I find extremely practical as a diabetic, I was praising its qualities again this morning to my diabetologist).

**Reply by j-kaltes** [reply]
Yes, you have to be careful about the labels you use because you have to use them your whole life. That keeps you alert.

**Reply by heurist1** [reply]
Hi Jaap

I have a prototype solution for notes, that still requires some tidying and testing.  However you obviously know so much about your software and the potential pit-falls, it worries me that I have forgotten something fundamental.

You said that it is not possible to do free text notes. Was that that it is not possible with current versions, or that there is a fundamental reason that it can't be done? (that's sort of the worry)

So far I have 
* made it possible to select which amount/number type, represents notes. In the same way that you do for meals. 
* Modified the editor so that you can add a string rather than a value, when this type is selected. Again I followed some of the ethos of meals.
* Keeping the notes in a separate storage, with a reference value in the amount.
* Limit the note to the first 10 chars when displayed on the graph, with an indication if they are longer.
* Added the syncing of notes, and a way to select if notes are too be synced. 
* Uploading of notes to LibreView
* Automatic purging of old data, so the storage can't get too large.

I've been running it on my device, linked to a Libre2 for a few days now, and it all appears to work ok. I still have to check how it looks from the LibreView end (I'll set up a second account to do this), and complete my code review as I want the change to be tidy and constrained. I have gone out of my way not to change things unnecessarily, and have done things like borrowing a byte from an unused/reserved field in settings to not change the structure size.

I will commit the changes to my fork soon. If I was to create a pull request for the feature, should it have plenty of comment to explain all the new variables and functions, or the same level as the current code to make it blend in?

Cheers 
Simon

**Reply by heurist1** [reply]
Have now created a pull request for this feature, feel free to comment on the PR.  Even better if you are able to build the modified app, any testing would be very much appreciated.
#472
